### DISCUSSION TITLE
Rosetta Drone and AI

### DISCUSSION THREAD

**Discussion by stefansommer** [RosettaDrone/rosettadrone#133] [General]
In the rosettadrone_mini2 repository, we have had some discussions regarding interfacing machine learning/AI algorithms with rosetta drone https://github.com/m4xw/rosettadrone_mini2/discussions/19 Still in an early stage, but we thought that it could make sense to continue the discussion in the upstream repo as people are looking into it here and it may be of general interest.

From my side: We (four cs students and I) are looking into monocular depth estimation using neural networks (convolutional and vision transformers). I view this as a first step in more actual autonomous control. Next could e.g. be direct prediction of flight paths from camera input similar to https://www.science.org/doi/10.1126/scirobotics.abg5810 We are currently investigating this from different perspectives: convolutional vs. ViT, supervising with simulated data and transfer learning (between data sets or simulated to real data), unsupervised learning (consistency of depth prediction when camera position/orientation changes, downscaled inference networks for running on device, etc.

We are currently using Rosetta Drone and dronekit to get telemetry data and then coupling this with saved video streams for training and evaluating models offline. It would be great to get this to work online on the UDP videostream, but so far we haven't managed to get the streaming to an external computer to work (it is a little while since we tested it, it may be that it works now). The depth estimation networks do work in real time when tested on e.g. a tello drone.

This was just to start the discussion. It would be great to hear what you are working on, other projects, or perhaps ideas to improve the above projects.

**Comment by kripper** [comment]
Hi @stefansommer,

Welcome to RossetaDrone. We are on the same page.

1) I just released for you my RawVideoSender plugin, which I use to send raw frames with no video re-encoding, no video buffer latencies and maximum quality (https://github.com/The1only/rosettadrone/commit/7693a4142b72c766253f51e1b0ef0f4a70c44c9d). You will need to enable this plugin (https://github.com/kripper/rosettadrone/commit/c234b1c7b9372f85315dc86d970e1a386a48a1e2). To receive the frames [check this](https://github.com/kripper/vision-landing-2/blob/fc32bd8c0a51a3403037cd7fba46c663e72d1712/src/track_targets.cpp#L119).
But if you want to stream video, please check [this](https://github.com/The1only/rosettadrone/issues/131).

2) Our friend Andre van Calster (@avncalst) did some nice AI scritps time ago.
Check his videos here: https://www.youtube.com/watch?v=fEGiTKeYNMw
And here is his work: https://github.com/avncalst/drone_cnn/tree/master/rosetta

3) I'm currently working on Vision Landing:
https://github.com/kripper/vision-landing-2
One problem we both will have to address is described here: https://github.com/The1only/rosettadrone/issues/132

4) Also check in the issues section. There is a lot of information about using DroneKit with Rosetta.

**Reply by stefansommer** [reply]
Excellent, thank you! We will try both streaming methods. I agree about https://github.com/The1only/rosettadrone/issues/132. In principle, the depth estimation could give a notion of height but how precise it will be is to be evaluated and point 2. will apply, thus the need for EKF

**Reply by OMAHDrone** [reply]
Dear Prof.Stefan,
May you help us how to setup dronekit + dji mini 2 with video streaming and control?
I tried to use rossetadrone3 app but l could not connect with dronekit nor l could control the dji mini 2 using rosettadrone3 app.
Thank you very much for your help
My pleasure

**Reply by kripper** [reply]
> Excellent, thank you! We will try both streaming methods. I agree about #132. In principle, the depth estimation could give a notion of height but how precise it will be is to be evaluated and point 2. will apply, thus the need for EKF

Hi @stefansommer, I'm currently doing successful tests with vision based precision landing.
I just added the timestamp and the yaw to the `RawVideoStreamer` packets to deal with video latency.
For position estimation (lat/lng), the flight controller data is useless. Only yaw and altitude are usefull. For the rest, I'm using computer vision:

![WhatsApp Image 2023-04-04 at 22 15 36](https://user-images.githubusercontent.com/1479804/230237192-837161ba-abf5-405d-8dde-7da71cb89408.jpg)

Please let me know about your progress.
If you have a github repo, I may also collaborate.

I have also been developing a framework for measuring the drone dynamics using the pose estimation as a reference. It's like a laboratory (an environment for testing algorithms and dynamics) that executes a series of tests to measures latencies, effective velocities, stability, generates graphs in real time, etc.

**Comment by kripper** [comment]
Please note that the EKF is needed mostly for dealing with the delay between the time you take the image and the time you finished computing and are able use these estimations (always too late).
So instead of using computer vision calculations and positioning sensors (which also have some a delay), you probably want to use a model to 1) estimate the current position (based on the EKF model) and 2) predict the future position (to be able to react on time especially when dealing with moving objects).

**Comment by kripper** [comment]
My first landing test:

[![First basic test](https://img.youtube.com/vi/tOXuLmG5JBc/0.jpg)](https://www.youtube.com/watch?v=tOXuLmG5JBc)

**Reply by OMAHDrone** [reply]
Thank you very much kripper for the video and it is great
May you help us with the setup. We need only the minimum dronekit code that will receive the video feed and do some very basic control. We are working on deep reinforcement learning project using dji mini 2. May you help us with setup. We can collaborate with you and update you.
Thank you very much again for your help and time

**Reply by kripper** [reply]
> Thank you very much kripper for the video and it is great May you help us with the setup. We need only the minimum dronekit code that will receive the video feed and do some very basic control. We are working on deep reinforcement learning project using dji mini 2. May you help us with setup. We can collaborate with you and update you. Thank you very much again for your help and time

Yes. I will help you both with your AI projects.
I gave the instructions in my first comment here, but I will create a detailed setup guide for you and release my latest versions.

**Reply by stefansommer** [reply]
very nice video! Great to hear about the progress, and the dynamics testing setup.
So far, we have mostly worked on the different neural network designs for depth estimation, trained on existing datasets and simulated data. No so much progress yet running it on live data. 
@OMAHDrone , do you also have problems with the dronekit connection? For us, video has been an issue, but dronekit has worked relatively well.

**Reply by OMAHDrone** [reply]
Thank you very much kripper  and Prof.Stefan
Yes l have problems with dronekit connection. I could not mange to communicate with the python dronekit and dji mini2
May you help us Prof.Stefan with the following regrading dji mini2:
1- We need the rosettadrone app you use to communicate with dji mini2.
2- We need the basic dronekit python code you used to receive the video -if possible - and do some basic motion control
We can collaborate with you Prof.Stefan and kripper and make our projects open source. I have to submit my project in June. That is why I am in hurry to quickly finish the setup and develop the AI network because of time constrains. Basically what l am trying to do is to develop an AI agent that will enable dji mini2 to track and land over a moving car using deep reinforcement learning.
The basic dronekit python code that we need is like this:
.....
takeoff
While true:
   receive drone image
   get drone states like speed and altitude (if possible)
   send these data to AI agent
   let AI agent generate control signals
  send control signals back to drone
  Repeat until mission is complete
  then land
...
We are ready to collaborate with you kripper and Prof.Stefan and make our projects open source to help the community and researchers
We are waiting for any instructions form you to start
Thank you very much again for your help and time
My pleasure

**Comment by kripper** [comment]
Here are the instructions (draft):
https://docs.google.com/document/d/1wR56Mvd2OsT82kxMOD7f_3-c6yJGBtLaT8LuVrmC5wI/edit#heading=h.q9mp6yjsfxe5

**Reply by OMAHDrone** [reply]
Thank you very much kripper for the instructions 
I am currently reading it and l will reply back after testing
Thank you very much for your help and support
My pleasure

**Reply by OMAHDrone** [reply]
Dear kripper, I have tested the code.
There is a problem when I recieve the video from dji mini 2. I did the following:
1- I have modified the app as suggested in the instruction document and built it. Basically, I have enabled the plugin and modified the host name and port number.
2- I have changed the ip of my Ubuntu laptop to be the same like the one set in the app.
3- I tried to receive the video directly in python2.7 on Ubuntu 16 using the given sample code.
4- I got the following error in smart_landing -- rosetta_camera.py, which has to do with reshaping the received image:
ValueError: cannot reshape array of size  1384448 into shape (1080, 1280)
in:  yuv_mat = yuv_mat.reshape((h + h // 2, w))
It is noted that the connection between the app and python video receiving script is only done when the app is in the Test mode. The connection is not working when the app is in drone or simulator mode.
May you help me with debugging this?
Thank you very much for your help
My pleasure

**Comment by kripper** [comment]
Are you 100% sure you pulled the latest Rosetta changes?

**Reply by OMAHDrone** [reply]
Thank you very much kripper
I have performed this operation on 7 april:
git clone https://github.com/RosettaDrone/rosettadrone
Does this ensure that I have the latest version? or do l have to perform another steps. If yes, how should I write the command.
Thank you very much for your help and support
My pleasure

**Reply by kripper** [reply]
> Thank you very much kripper I have performed this operation on 7 april: git clone https://github.com/RosettaDrone/rosettadrone Does this ensure that I have the latest version? or do l have to perform another steps. If yes, how should I write the command. Thank you very much for your help and support My pleasure

Just pull.

**Comment by kripper** [comment]
The method sendYuvData is receiving 2048 additional bytes in your case. Maybe a header. Try to skip the first or last 2048 bytes of the buffer and set size = 1080 * 1280 so you can see something.

**Reply by OMAHDrone** [reply]
Thank you very much kripper
I will test it and update you as soon as I finish it
Thank you very much for your help
My pleasure

**Reply by OMAHDrone** [reply]
Dear kripper,
I have done the following:
1- Went to inside the downloaded repository on my local PC and performed : git pull. Then I re-built the app.
2- I modified the buffer as buffer[2048:] and now l can receive the video. But the video is very slow, glitches, sometimes with color and sometimes gray
3- The yaw reading is always zero. But when l change to drone mode, it updates. But I can not start the app in Drone mode. I have first to start the app in Test mode and then go back to the main menu and try several times until Drone mode is activated. This is because when  -in some cases - I open Drone mode, the phone screen starts to rotate as if it will open the Drone mode but it redo the rotation and returns back to the main menu.
May l know why the update rate of this video is very slow and part of the received images seams to be not from the same scene?
My setup is like this:
- Ubuntu 16 , 8 RAM, intel core i5, intel HD graphics, dji mini 2
-  I use twp phones, one phone has the app installed, another phone makes a public wifi network, then the PC and app phone connect to the same network.
Thank you very much for your help
My pleasure

**Reply by tr0043t** [reply]
Android MSDK 5.3.0 with Mini 3/Pro support is now up.

https://developer.dji.com/

**Reply by kripper** [reply]
> the video is very slow

What is slow? It's configured for 15 fps and it's uncompressed. You can increase it, but how much fps is your algorithm going to process?

> glitches, sometimes with color and sometimes gray

You are omitting 2048 bytes and we don't know what format you are receiving.
Or you didn't place a fan in front and it's overheating.

> 3- The yaw reading is always zero. But when l change to drone mode, it updates.

Test mode doesn't send real telemetry.

> But I can not start the app in Drone mode. I have first to start the app in Test mode and then go back to the main menu and try several times until Drone mode is activated.

Maybe a problem with the cable or your smartphone or something that prevents proper detection. 

> in some cases - I open Drone mode, the phone screen starts to rotate as if it will open the Drone mode but it redo the rotation and returns back to the main menu.

Maybe the activity is finishing because the flight controller is null (not correctly detected).
Check the logs. Use breakpoints.

You should also first test and comment this:
https://github.com/The1only/rosettadrone/issues/131

And for reporting Issues, create issues, one issue per issue.

**Reply by kripper** [reply]
> But I can not start the app in Drone mode

Issue created here: https://github.com/RosettaDrone/rosettadrone/issues/152

**Reply by OMAHDrone** [reply]
Thank you very much kripper
- Q1) What is slow? It's configured for 15 fps and it's uncompressed. You can increase it, but how much fps is your algorithm going to process.
 I am suffering from very high latency which can reach up to 30 seconds or more. When I decrease the fps to 1, the latency reduces to about 10 seconds. I also noted that when I change the w and h variables in the app both of them to 1 instead of (1280, 720), the latency decreases but still the received image is not normal. What I mean is that the received image is like taking the true image and then dividing it into two halves and then flipping the two halves and form the "the received image". It is not a true normal representation of the scene.

- Q2) You are omitting 2048 bytes and we don't know what format you are receiving.
Or you didn't place a fan in front and it's overheating.
  How to change format then? I am using an air condition in front of the drone.
- Q3) Test mode doesn't send real telemetry.
  ok noted.
- Q4) Maybe a problem with the cable or your smartphone or something that prevents proper detection. Maybe the activity is finishing because the flight controller is null (not correctly detected). Check the logs. Use breakpoints.
 I have re-downloaded the repository and it works great now.
- Q5) You should also first test and comment this:
https://github.com/RosettaDrone/rosettadrone/issues/131
And for reporting Issues, create issues, one issue per issue.
- I will check it
Thank you very much kripper for your help and time
My pleasure

**Reply by kripper** [reply]
Ok. I'm hidding this discussion. Make sure to report issues correctly.
